lms/djangoapps/bulk_email/tests/test_tasks.py

```diff
@@ -278,6 +278,26 @@ def test_ses_domain_ends_with_dot(self):
         # Test that celery handles permanent SMTPDataErrors by failing and not retrying.
         self._test_email_address_failures(SESDomainEndsWithDotError(554, "Email address ends with a dot"))
 
+    def test_email_address_with_ascii_characters_failed(self):
+        """
+        Test that celery handles with ascii characters in email address and
+        failed the email and not retrying."""
+        num_emails = 10
+        # We also send email to the instructor:
+        students = [self.create_student('robot%d' % i) for i in xrange(num_emails - 1)]
+        for i in xrange(len(students)):
+            students[i].email = ('robot%d@tesá.com' % i)
+            students[i].save()
+
+        expected_fails = num_emails - 1
+        expected_succeeds = num_emails - expected_fails
+
+        with patch('bulk_email.tasks.get_connection', autospec=True) as get_conn:
+            get_conn.return_value.send_messages.side_effect = cycle([None])
```

Comment:
Not sure if this a good idea to patch `get_connection` here and raise an exception. Do you know a good reason why its been done here?

Reply:
Good question.  Actually, I want to patch a function for only some part of a test. I do not use the patch as a context manager when I want a function patched for an entire test.

Reply:
@attiyaIshaque  I took a look at this PR and just made the test more readable. Please check if this works for you? Maybe you can have some bits of it. 
```
    def test_bulk_email_skip_with_non_ascii_emails(self):
        """
        Tests that bulk email skips the email address containing non-ASCII characters
        and does not fail.
        """
        num_emails = 10
        emails_with_non_ascii_chars = 3
        numb_course_instructors = 1

        students = [self.create_student('robot%d' % i) for i in xrange(num_emails)]
        for index, student in enumerate(students, 1):
            if index > emails_with_non_ascii_chars:
                break
            student.email = ('robot%d@tesá.com' % index)
            student.save()

        total = num_emails + numb_course_instructors
        expected_succeeds = num_emails - emails_with_non_ascii_chars + numb_course_instructors
        expected_fails = emails_with_non_ascii_chars

        with patch('bulk_email.tasks.get_connection', autospec=True) as get_conn:
            get_conn.return_value.send_messages.side_effect = cycle([None])
            self._test_run_with_task(
                task_class=send_bulk_course_email,
                action_name='emailed',
                total=total,
                succeeded=expected_succeeds,
                failed=expected_fails
            )
```